File: 2015/21_RPGSimulator20XX/rpgsim.js
/* eslint-disable linebreak-style */
// ======================================================================
// RPG Simulator 20XX
//   Advent of Code 2015 Day 21 -- Eric Wastl -- https://adventofcode.com
//
// JavaScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           r p g s i m . j s
//
// A solver for the Advent of Code 2015 Day 21 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
const aoc21 = require('./aoc_21');
const shop = require('./shop');
const player = require('./player');
const battle = require('./battle');

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const standardShop = new shop.Shop({});
const standardPlayer = new player.Player({
  name: 'player',
  text: aoc21.fromText('Hit Points: 100\nDamage: 0\nArmor: 0'),
});
const maximumCost = 113;

// ======================================================================
//                                                                 RPGSim
// ======================================================================

class RPGSim {
  // Object for RPG Simulator 20XX

  constructor(options) {
    // Create a RPGSim object

    // 1. Set the initial values
    this.text = options.text === undefined ? null : options.text;
    this.part2 = options.part2 === undefined ? false : options.part2;
    this.shop = options.shop === undefined ? standardShop : options.shop;
    this.boss = options.boss === undefined ? null : options.boss;
    this.player = options.player === undefined ? standardPlayer : options.player;

    // 2. Process text (if any)
    if (this.text !== null) {
      this.boss = new player.Player({ part2: this.part2, name: 'boss', text: this.text });
    }

    // 3. Set up for the conflict
    this.battle = new battle.Battle({ part2: this.part2, player: this.player, boss: this.boss });
  }

  findCheapestVictory() {
    // 1. Start with a known (or assumed) value that is too high
    let bestCost = maximumCost;
    // 2. Get the keys to the weapons, armor, and rings
    const allKeys = this.shop.shopKeys();
    // 3. Loop for all of the weapons
    for (let windex = 0; windex < allKeys.weapons.length; windex += 1) {
      const wname = allKeys.weapons[windex];
      const weapon = this.shop.weapons[wname];
      // eslint-disable-next-line no-console
      // console.log(`weapon: ${wname} cost = ${weapon.cost} ${typeof weapon.cost}`);
      if (weapon.cost < bestCost) {
        // 4. Loop for all armor choices (including no armor)
        for (let aindex = 0; aindex < allKeys.xarmor.length; aindex += 1) {
          const aname = allKeys.xarmor[aindex];
          const armor = this.shop.xarmor[aname];
          // eslint-disable-next-line no-console
          // console.log(`armor: ${aname} cost = ${armor.cost}`);
          if ((weapon.cost + armor.cost) < bestCost) {
            // 5. Loop for all the ring choices (include none, one, or two)
            for (let rindex = 0; rindex < allKeys.xrings.length; rindex += 1) {
              const rname = allKeys.xrings[rindex];
              const rings = this.shop.xrings[rname];
              // eslint-disable-next-line no-console
              // console.log(`rings: ${rname} cost = ${rings.cost}`);
              const totalCost = weapon.cost + armor.cost + rings.cost;
              if (totalCost < bestCost) {
                // 6. Set the damage and armor attributes of the player
                this.player.damage = weapon.damage + armor.damage + rings.damage;
                this.player.armor = weapon.armor + armor.armor + rings.armor;
                // 7. Send the player into battle
                const victor = this.battle.simulate();
                // 8. Did he win?
                if (victor.name === this.player.name) {
                  bestCost = totalCost;
                  // eslint-disable-next-line no-console
                  // console.log(`New best cost of ${bestCost}: w:${weapon.name}, a:${armor.name}, r:${rings.name}, hp=${victor.hitPoints}`);
                }
              }
            }
          }
        }
      }
    }
    // 9. Return the best cost
    return bestCost;
  }

  findCostliestDefeat() {
    // 1. Start with a known to bee too low
    let bestCost = -1;
    // 2. Get the keys to the weapons, armor, and rings
    const allKeys = this.shop.shopKeys();
    // 3. Loop for all of the weapons
    for (let windex = 0; windex < allKeys.weapons.length; windex += 1) {
      const wname = allKeys.weapons[windex];
      const weapon = this.shop.weapons[wname];
      // eslint-disable-next-line no-console
      // console.log(`weapon: ${wname} cost = ${weapon.cost} ${typeof weapon.cost}`);
      // 4. Loop for all armor choices (including no armor)
      for (let aindex = 0; aindex < allKeys.xarmor.length; aindex += 1) {
        const aname = allKeys.xarmor[aindex];
        const armor = this.shop.xarmor[aname];
        // eslint-disable-next-line no-console
        // console.log(`armor: ${aname} cost = ${armor.cost}`);
        // 5. Loop for all the ring choices (include none, one, or two)
        for (let rindex = 0; rindex < allKeys.xrings.length; rindex += 1) {
          const rname = allKeys.xrings[rindex];
          const rings = this.shop.xrings[rname];
          // eslint-disable-next-line no-console
          // console.log(`rings: ${rname} cost = ${rings.cost}`);
          const totalCost = weapon.cost + armor.cost + rings.cost;
          if (totalCost > bestCost) {
            // 6. Set the damage and armor attributes of the player
            this.player.damage = weapon.damage + armor.damage + rings.damage;
            this.player.armor = weapon.armor + armor.armor + rings.armor;
            // 7. Send the player into battle
            const victor = this.battle.simulate();
            // 8. Did he win and was it cheaper?
            if (victor.name !== this.player.name) {
              bestCost = totalCost;
              // eslint-disable-next-line no-console
              console.log(`New higher cost of ${bestCost}: w:${weapon.name}, a:${armor.name}, r:${rings.name}, hp=${victor.hitPoints}`);
            }
          }
        }
      }
    }
    // 9. Return the best cost
    return bestCost;
  }

  partOne(options) {
    // Returns the solution for part one

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part one
    return this.findCheapestVictory();
  }

  partTwo(options) {
    // Returns the solution for part two

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part two
    return this.findCostliestDefeat();
  }
}

module.exports.RPGSim = RPGSim;
// ======================================================================
// end                        r p g s i m . j s                       end
// ======================================================================

File: 2015/21_RPGSimulator20XX/rpgsim.py
# ======================================================================
# RPG Simulator 20XX
#   Advent of Code 2015 Day 21 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         r p g s i m . p y
# ======================================================================
"A solver for the Advent of Code 2015 Day 21 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import player
import shop

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                 Rpgsim
# ======================================================================


class Rpgsim(object):   # pylint: disable=R0902, R0205
    "Object for RPG Simulator 20XX"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.shop = shop.Shop()
        self.boss = None

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self.boss = player.Player(text)

    def boss_wins_fight(self, items):
        "Have the boss fight the player"

        # 1. Create the combatents
        boss = self.boss.clone()
        plyr = player.Player(part2=self.part2, items=items)

        # 2. Do battle until one or the other is dean
        while not (boss.is_dead() or plyr.is_dead()):

            # 3. Player attacks
            if boss.defend(plyr.attack()):
                return False

            # 4. Boss attacks
            if plyr.defend(boss.attack()):
                return True

    def cheapest_win(self, verbose=False):
        "Determine the cheapeast win"

        # 1. Start with nothing
        cheapest_cost = 999999999999
        cheapest_items = []

        # 2. Loop for all the possible items for the player
        for items in self.shop.combinations():

            # 3. Battle
            if not self.boss_wins_fight(items):

                # 4. Player won, is this win cheaper?
                this_cost = shop.Shop.total_price(items)
                if this_cost < cheapest_cost:
                    cheapest_cost = this_cost
                    cheapest_items = items
                    if verbose:
                        print('Cheaper %d %s' % (cheapest_cost, shop.Shop.item_names(items)))

        # 5. Return the cheapest win cost
        if verbose:
            print('Cheapest Win $%d with %s' %
                  (cheapest_cost, shop.Shop.item_names(cheapest_items)))
        return cheapest_cost

    def costliest_loss(self, verbose=False):
        "Determine the the most expensive loss"

        # 1. Start with nothing
        expensive_cost = -1
        expensive_items = []

        # 2. Loop for all the possible items for the player
        for items in self.shop.combinations():

            # 3. Battle
            if self.boss_wins_fight(items):

                # 4. boss won, is this loss more expensive?
                this_cost = shop.Shop.total_price(items)
                if this_cost > expensive_cost:
                    expensive_cost = this_cost
                    expensive_items = items
                    if verbose:
                        print('Costlier %d %s' % (expensive_cost, shop.Shop.item_names(items)))

        # 5. Return the most expensive loss win cost
        if verbose:
            print('Most Expensive $%d with %s' %
                  (expensive_cost, shop.Shop.item_names(expensive_items)))
        return expensive_cost

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.cheapest_win(verbose=verbose)

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.costliest_loss(verbose=verbose)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        r p g s i m . p y                       end
# ======================================================================
